lib/ddtrace/contrib/faraday/middleware.rb

```diff
@@ -0,0 +1,75 @@
+require 'faraday'
+require 'ddtrace/ext/http'
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+
+module Datadog
+  module Contrib
+    module Faraday
+      # Middleware implements a faraday-middleware for ddtrace instrumentation
+      class Middleware < ::Faraday::Middleware
+        DEFAULT_ERROR_HANDLER = lambda do |env|
+          Ext::HTTP::ERROR_RANGE.cover?(env[:status])
+        end
+
+        DEFAULT_OPTIONS = {
+          distributed_tracing: false,
+          split_by_domain: false,
+          error_handler: DEFAULT_ERROR_HANDLER
+        }.freeze
+
+        def initialize(app, options = {})
+          super(app)
+          @options = DEFAULT_OPTIONS.merge(options)
+        end
+
+        def call(env)
+          dd_pin.tracer.trace(SERVICE) do |span|
+            annotate!(span, env)
+            propagate!(span, env) if options[:distributed_tracing]
+            app.call(env).on_complete { |resp| handle_response(span, resp) }
+          end
+        end
+
+        private
+
+        attr_reader :app, :options
+
+        def annotate!(span, env)
+          span.resource = resource_name(env)
+          span.service = dd_pin.service
+          span.span_type = Ext::HTTP::TYPE
+          span.set_tag(Ext::HTTP::URL, env[:url].path)
+          span.set_tag(Ext::HTTP::METHOD, env[:method].upcase)
+          span.set_tag(Ext::NET::TARGET_HOST, env[:url].host)
+          span.set_tag(Ext::NET::TARGET_PORT, env[:url].port)
+        end
+
+        def handle_response(span, env)
+          if options.fetch(:error_handler).call(env)
+            span.set_error(["Error #{env[:status]}", env[:body]])
+          end
+
+          span.set_tag(Ext::HTTP::STATUS_CODE, env[:status])
+        end
+
+        def propagate!(span, env)
+          env[:request_headers].merge!(
+            Ext::DistributedTracing::HTTP_HEADER_TRACE_ID => span.trace_id,
+            Ext::DistributedTracing::HTTP_HEADER_PARENT_ID => span.parent_id
+          )
+        end
+
+        def dd_pin
+          Pin.get_from(::Faraday)
+        end
+
+        def resource_name(env)
```

Comment:
oh it's the `service` that should change the name, not the `resource` otherwise we may get high cardinality in our list. It should follow the approach described here: https://github.com/DataDog/dd-trace-py/pull/252/files#r138045288